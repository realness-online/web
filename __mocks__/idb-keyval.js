module.exports.get = jest.fn(() => Promise.resolve(undefined))
module.exports.set = jest.fn(() => Promise.resolve())
module.exports.del = jest.fn(() => Promise.resolve())
module.exports.clear = jest.fn(() => Promise.resolve())
module.exports.keys = jest.fn(() => Promise.resolve([]))
