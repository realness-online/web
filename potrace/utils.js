/**
 * @typedef {import('@/potrace/types/Point').default} Point
 */

import Point from '@/potrace/types/Point'

const attr_regexps = {}

/**
 * @param {string} attr_name
 * @returns {RegExp}
 */
const get_attr_regexp = attr_name => {
  if (attr_regexps[attr_name]) {
    return attr_regexps[attr_name]
  }

  attr_regexps[attr_name] = new RegExp(
    ` ${attr_name}="((?:\\\\(?=")"|[^"])+)"`,
    'i'
  )
  return attr_regexps[attr_name]
}

/**
 * @param {string} html
 * @param {string} attr_name
 * @param {string} value
 * @returns {string}
 */
export const set_html_attr = (html, attr_name, value) => {
  const attr = ` ${attr_name}="${value}"`

  if (html.indexOf(` ${attr_name}="`) === -1) {
    html = html.replace(/<[a-z]+/i, beginning => beginning + attr)
  } else {
    html = html.replace(get_attr_regexp(attr_name), attr)
  }

  return html
}

/**
 * @param {number} number
 * @returns {string}
 */
const fixed = number => number.toFixed(3).replace('.000', '')

/**
 * @param {number} a
 * @param {number} n
 * @returns {number}
 */
export const mod = (a, n) =>
  a >= n ? a % n : a >= 0 ? a : n - 1 - ((-1 - a) % n)

/**
 * @param {Point} p1
 * @param {Point} p2
 * @returns {number}
 */
export const xprod = (p1, p2) => p1.x * p2.y - p1.y * p2.x

/**
 * @param {number} a
 * @param {number} b
 * @param {number} c
 * @returns {boolean}
 */
export const cyclic = (a, b, c) => (a <= c ? a <= b && b < c : a <= b || b < c)

/**
 * @param {number} i
 * @returns {-1 | 0 | 1}
 */
export const sign = i => (i > 0 ? 1 : i < 0 ? -1 : 0)

/**
 * @param {Object} Q Matrix-like object with at(i,j) method
 * @param {Point} w
 * @returns {number}
 */
export const quadform = (Q, w) => {
  const v = [w.x, w.y, 1]
  let sum = 0

  for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
      sum += v[i] * Q.at(i, j) * v[j]
    }
  }
  return sum
}

/**
 * @param {number} lambda
 * @param {Point} a
 * @param {Point} b
 * @returns {Point}
 */
export const interval = (lambda, a, b) => {
  const res = new Point()
  res.x = a.x + lambda * (b.x - a.x)
  res.y = a.y + lambda * (b.y - a.y)
  return res
}

/**
 * @param {Point} p0
 * @param {Point} p2
 * @returns {Point}
 */
export const dorth_infinity = (p0, p2) => {
  const r = new Point()
  r.y = sign(p2.x - p0.x)
  r.x = -sign(p2.y - p0.y)
  return r
}

export const ddenom = (p0, p2) => {
  const r = dorth_infinity(p0, p2)
  return r.y * (p2.x - p0.x) - r.x * (p2.y - p0.y)
}

export const dpara = (p0, p1, p2) => {
  const x1 = p1.x - p0.x
  const y1 = p1.y - p0.y
  const x2 = p2.x - p0.x
  const y2 = p2.y - p0.y

  return x1 * y2 - x2 * y1
}

export const cprod = (p0, p1, p2, p3) => {
  const x1 = p1.x - p0.x
  const y1 = p1.y - p0.y
  const x2 = p3.x - p2.x
  const y2 = p3.y - p2.y

  return x1 * y2 - x2 * y1
}

export const iprod = (p0, p1, p2) => {
  const x1 = p1.x - p0.x
  const y1 = p1.y - p0.y
  const x2 = p2.x - p0.x
  const y2 = p2.y - p0.y

  return x1 * x2 + y1 * y2
}

export const iprod1 = (p0, p1, p2, p3) => {
  const x1 = p1.x - p0.x
  const y1 = p1.y - p0.y
  const x2 = p3.x - p2.x
  const y2 = p3.y - p2.y

  return x1 * x2 + y1 * y2
}

export const ddist = (p, q) =>
  Math.sqrt((p.x - q.x) * (p.x - q.x) + (p.y - q.y) * (p.y - q.y))

export const luminance = (r, g, b) =>
  Math.round(0.2126 * r + 0.7153 * g + 0.0721 * b)

export const between = (val, min, max) => val >= min && val <= max

export const clamp = (val, min, max) => Math.min(max, Math.max(min, val))

export const is_number = val => typeof val === 'number'

/**
 * @typedef {Object} Curve
 * @property {Point[]} c Array of curve points
 * @property {number} n Number of segments
 * @property {('CURVE'|'CORNER')[]} tag Tags for each segment
 */

/**
 * @param {Curve} curve
 * @param {number} [scale=1]
 * @returns {string} SVG path data
 */
export const render_curve = (curve, scale = 1) => {
  const starting_point = curve.c[(curve.n - 1) * 3 + 2]
  let path = `M ${fixed(starting_point.x * scale)} ${fixed(starting_point.y * scale)} `

  curve.tag.forEach((tag, i) => {
    const i3 = i * 3
    const p0 = curve.c[i3]
    const p1 = curve.c[i3 + 1]
    const p2 = curve.c[i3 + 2]

    if (tag === 'CURVE') {
      path += `C ${fixed(p0.x * scale)} ${fixed(p0.y * scale)}, `
      path += `${fixed(p1.x * scale)} ${fixed(p1.y * scale)}, `
      path += `${fixed(p2.x * scale)} ${fixed(p2.y * scale)} `
    } else if (tag === 'CORNER') {
      path += `L ${fixed(p1.x * scale)} ${fixed(p1.y * scale)} `
      path += `${fixed(p2.x * scale)} ${fixed(p2.y * scale)} `
    }
  })

  return path
}

/**
 * @param {number} t Parameter between 0 and 1
 * @param {Point} p0 Start point
 * @param {Point} p1 First control point
 * @param {Point} p2 Second control point
 * @param {Point} p3 End point
 * @returns {Point}
 */
export const bezier = (t, p0, p1, p2, p3) => {
  const s = 1 - t
  const res = new Point()

  res.x =
    s * s * s * p0.x +
    3 * (s * s * t) * p1.x +
    3 * (t * t * s) * p2.x +
    t * t * t * p3.x
  res.y =
    s * s * s * p0.y +
    3 * (s * s * t) * p1.y +
    3 * (t * t * s) * p2.y +
    t * t * t * p3.y

  return res
}

/**
 * @param {Point} p0
 * @param {Point} p1
 * @param {Point} p2
 * @param {Point} p3
 * @param {Point} q0
 * @param {Point} q1
 * @returns {number}
 */
export const tangent = (p0, p1, p2, p3, q0, q1) => {
  var A, B, C, a, b, c, d, s, r1, r2

  A = cprod(p0, p1, q0, q1)
  B = cprod(p1, p2, q0, q1)
  C = cprod(p2, p3, q0, q1)

  a = A - 2 * B + C
  b = -2 * A + 2 * B
  c = A

  d = b * b - 4 * a * c

  if (a === 0 || d < 0) return -1.0

  s = Math.sqrt(d)

  r1 = (-b + s) / (2 * a)
  r2 = (-b - s) / (2 * a)

  if (r1 >= 0 && r1 <= 1) return r1
  if (r2 >= 0 && r2 <= 1) return r2
  return -1.0
}

export default {
  luminance,
  between,
  clamp,
  is_number,
  set_html_attr,
  render_curve,
  bezier,
  tangent,
  mod,
  xprod,
  cyclic,
  sign,
  quadform,
  interval,
  ddenom,
  dpara,
  cprod,
  iprod,
  iprod1,
  ddist,
  dorth_infinity
}
