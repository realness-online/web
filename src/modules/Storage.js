import Item from '@/modules/Item'
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'

class Storage {
  static hydrate(item_as_string) {
    return document.createRange().createContextualFragment(item_as_string)
  }
  constructor(item_type, selector = `[itemtype="/${item_type}"]`,
    filename = `${item_type}.html`) {
    this.item_type = item_type
    this.selector = selector
    this.filename = filename
    this.metadata = {'contentType': 'text/html'}
  }
  from_storage() {
    let storage_string = localStorage.getItem(this.item_type)
    return Storage.hydrate(storage_string)
  }
  as_list() {
    return Item.get_items(this.from_storage())
  }
  as_object() {
    return Item.get_first_item(this.from_storage())
  }
  save() {
    // save the information to local storage and if appropriate to the server
    let items = document.querySelector(this.selector)
    if (!items) { return false }
    items = items.outerHTML
    localStorage.setItem(this.item_type, items)
    if (['person', 'posts'].includes(this.item_type)) {
      this.persist(items)
    }
    return true
  }
  sync() {
    // check if an 'item_type'.html already exist for this user.
    // If so merge it's content whith what's stored locally
    const url = this.get_server_file_url()
    if (['person', 'posts'].includes(this.item_type) && navigator.onLine) {
      firebase.auth().onAuthStateChanged(user => {
        if (user) {
          const doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
          firebase.storage().ref().child(doc_u_path).getDownloadURL().then((url) => {
            fetch(url).then(response => {
              response.text().then(server_text => {
                const server_as_fragment = Storage.hydrate(server_text)
                let server_as_list = Item.get_items(server_as_fragment)
                console.log(server_as_list)
              })
              let items = [...this.from_storage, ...from_server]
              // console.log(items)
            })
          })
        }
      })
    }
  }

  get_server_file_url() {

  }

  persist(doc_u_ment, doc_u_path) {
    firebase.auth().onAuthStateChanged(user => {
      if (user && navigator.onLine) {
        const file = new File([doc_u_ment], this.filename)
        if (!doc_u_path) {
          doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
        }
        firebase.storage().ref()
          .child(doc_u_path)
          .put(file, this.metadata)
          .catch(console.log.bind(console))
      }
    })
  }
}

export default Storage
export const person_storage = new Storage('person')
export const posts_storage = new Storage('posts', '[itemprop=posts]')
export const activity_storage = new Storage('activity', '[itemprop=activity]')
export const relations = new Storage('relations', '[itemprop=relations]')
export const phonebook = new Storage('phonebook')

//  use this for firebase logging
// .then( snapshot => {
//   console.log('Uploaded', snapshot.totalBytes, 'bytes.')
//   console.log(snapshot.metadata)
//   console.log(snapshot.fullPath)
//   console.log('File available at', snapshot.downloadURL)
// })
