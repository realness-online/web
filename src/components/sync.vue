<template lang="html">
  <div ref="sync" hidden>
    <as-days v-if="show_statements" v-slot="thoughts"
             itemscope
             :itemid="itemid('statements')"
             :statements="statements">
      <thought-as-article v-for="thought in thoughts"
                          :key="thought[0].id"
                          :statements="thought" />
    </as-days>
    <events-list v-if="events" :events="events" :itemid="itemid('events')" />
    <unsynced-posters v-for="poster in posters"
                      :key="poster.id"
                      :immediate="true"
                      :itemid="poster.id"
                      :poster="poster" />
  </div>
</template>
<script>
  import * as firebase from 'firebase/app'
  import 'firebase/auth'
  import hash from '@/modules/hash'
  import { set, get, del } from 'idb-keyval'
  import { as_type, as_created_at, list } from '@/helpers/itemid'
  import get_item from '@/modules/item'
  import { from_e64 } from '@/helpers/profile'
  import as_days from '@/components/as-days'
  import as_list from '@/components/events/as-list'
  import as_svg from '@/components/posters/as-svg'
  import thought_as_article from '@/components/statements/as-article'
  import {
    Events,
    Statements,
    Poster,
    Offline
  } from '@/persistance/Storage'
  export default {
    components: {
      'as-days': as_days,
      'events-list': as_list,
      'thought-as-article': thought_as_article,
      'unsynced-posters': as_svg
    },
    props: {
      statement: {
        type: Object,
        required: false,
        default: null
      }
    },
    data () {
      return {
        syncer: new Worker('/sync.worker.js'),
        syncing: false,
        person: null,
        posters: null,
        statements: [],
        events: null
      }
    },
    computed: {
      show_statements () {
        if (this.statement) return true
        else if (this.syncing) return true
        return false
      }
    },
    watch: {
      statement () {
        if (this.statement) this.save_statement()
      }
    },
    mounted () {
      firebase.auth().onAuthStateChanged(this.sync_local_storage)
      window.addEventListener('online', this.online)
    },
    beforeDestroy () {
      this.syncer.terminate()
    },
    methods: {
      async save_statement () {
        const itemid = this.itemid('statements')
        this.statements = await list(itemid)
        this.statements.push(this.statement)
        const data = new Statements()
        await this.$nextTick()
        await data.save()
        this.$emit('update:statement', null)
      },
      itemid (type) {
        if (type) return `${localStorage.me}/${type}`
        else return `${localStorage.me}`
      },
      online () {
        this.sync_local_storage(firebase.auth().currentUser)
      },
      async sync_local_storage (current_user) {
        if (navigator.onLine && current_user) {
          console.time('syncing:local-storage')
          this.syncing = true
          localStorage.me = from_e64(current_user.phoneNumber)
          // await Promise.all([
          //   this.sync_statements(),
          //   this.sync_offline(),
          //   this.sync_anonymous_posters(current_user),
          //   this.sync_events()
          // ])
          this.sync_offline()
          this.sync_anonymous_posters(current_user)
          this.sync_events()
          this.sync_statements()
          this.syncing = false
          console.timeEnd('syncing:local-storage')
        }
      },
      async sync_offline () {
        const offline = await get('offline')
        if (!offline) return
        while (offline.length) {
          const item = offline.pop()
          if (item.action === 'save') await new Offline(item.id).save()
          else if (item.action === 'delete') await new Offline(item.id).delete()
          else console.info('unknown offline action', item.action, item.id)
        }
        await del('offline')
      },
      async sync_anonymous_posters (my) {
        const offline_posters = await get('/+/posters/')
        if (!offline_posters || !offline_posters.items) return
        const posters = []
        await Promise.all(offline_posters.items.map(async (created_at) => {
          const poster_as_string = await get(`/+/posters/${created_at}`)
          const poster = get_item(poster_as_string)
          poster.id = `${from_e64(my.phoneNumber)}/posters/${created_at}`
          posters.push(poster)
        }))
        if (posters.length) {
          this.posters = posters
          await this.$nextTick()
          await Promise.all(this.posters.map(async (poster) => {
            const created_at = as_created_at(poster.id)
            const new_poster = new Poster(poster.id)
            await new_poster.save()
            await del(`/+/posters/${created_at}`)
            offline_posters.items = offline_posters.items.filter(when => {
              return parseInt(when) !== created_at
            })
            await set('/+/posters/', offline_posters)
          }))
          this.posters = []
        }
      },
      async sync_events () {
        const itemid = this.itemid('events')
        const events = new Events()
        this.events = await events.sync()
        await this.$nextTick()
        await this.sync_paged(itemid, events)
        this.events = null
      },
      async sync_statements () {
        const itemid = this.itemid('statements')
        const statements = new Statements()
        this.statements = await statements.sync()
        await this.$nextTick()
        await this.sync_paged(itemid, statements)
        this.statements = []
      },
      async sync_paged (itemid, paged) {
        const query = `[itemid="${itemid}"]`
        const elements = this.$el.querySelector(query)
        if (elements) {
          let index = await get('hash-index')
          if (!index) index = {}
          const current_hash = parseInt(index[itemid])
          const new_hash = hash(elements.outerHTML)
          if (current_hash !== new_hash) {
            await paged.save()
            localStorage.removeItem(`/+/${as_type(itemid)}`)
            index[itemid] = new_hash
            set('hash-index', index)
          }
        }
      }
    }
  }
</script>
