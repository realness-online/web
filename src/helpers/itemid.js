// {base_url}/{:author}/{:type}/{:created_at}
import firebase from 'firebase/app'
import 'firebase/storage'
import 'firebase/auth'
import get_item from '@/modules/item'
import { does_not_exist } from '@/persistance/Cloud.sync'
import { get, set } from 'idb-keyval'
// Expensive to call
export async function load(itemid, me = localStorage.me) {
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  item = await load_from_network(itemid, me)
  if (item) return item
  return null
}
export async function list(itemid, me = localStorage.me) {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    else return []
  } catch {
    return []
  }
}
export async function load_from_network(itemid, me = localStorage.me) {
  const url = await as_download_url(itemid, me)
  if (url) {
    console.info('download', itemid)
    const server_text = await (await fetch(url)).text()
    await set(itemid, server_text)
    return get_item(server_text)
  } else return null
}
export async function as_directory(itemid) {
  const path = as_directory_id(itemid)
  const cached = await get(path)
  if (cached) return cached
  else if (navigator.onLine && firebase.auth().currentUser) {
    const meta = { items: [], types: [] } // folders are types in our vocabulary
    console.info('request:directory', itemid)
    const directory = await firebase.storage().ref().child(`people/${path}`).listAll()
    directory.items.forEach(item => meta.items.push(item.name.split('.')[0]))
    directory.prefixes.forEach(prefix => meta.types.push(prefix.name))
    await set(path, meta)
    return meta
  } else return null
}
export async function as_download_url(itemid) {
  if (!firebase.auth().currentUser) return null
  const storage = firebase.storage().ref()
  try {
    console.info('request:location', itemid)
    return await storage.child(as_filename(itemid)).getDownloadURL()
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.warn(itemid, '=>', as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = does_not_exist
      await set('sync:index', index)
      return null
    } else throw e
  }
}

// Cheap to call
const created_at = ['avatars', 'posters', 'animations']
const has_history = ['statements', 'events']
export function is_history(itemid) {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  return false
}
export function as_filename(itemid) {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `/people${filename}`
  if (created_at.includes(as_type(itemid)) || is_history(itemid)) return `${filename}.html`
  else return `${filename}/index.html`
}
export function as_storage_path(itemid) {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}
export function as_directory_id(itemid) {
  const parts = as_path_parts(itemid)
  return `/${parts[0]}/${parts[1]}/`
}
export function as_path_parts(itemid) {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  return path
}
export function as_author(itemid) {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  else return null
}
export function as_type(itemid) {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  else return null
}
export function as_created_at(itemid) {
  const path = as_path_parts(itemid)
  return parseInt(path[2])
}
export function as_query_id(itemid) {
  return itemid.substring(2).replace('/', '-').replace('/', '-')
}
export function as_fragment(itemid) {
  return `#${as_query_id(itemid)}`
}
export function type_as_list(item) {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  else return []
}
