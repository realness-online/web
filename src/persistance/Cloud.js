// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'
import { download_url, as_filename } from '@/helpers/itemid'
import { del } from 'idb-keyval'
const networkable = ['person', 'statements', 'posters', 'avatars', 'events']
export const Cloud = (superclass) => class extends superclass {
  async to_network (items) {
    const user = firebase.auth().currentUser
    const path = as_filename(this.id)
    if (user && navigator.onLine) {
      const file = new File([items], path)
      await firebase.storage().ref().child(path).put(file, this.metadata)
    }
  }
  async save (items = document.querySelector(`[itemid="${this.id}"]`)) {
    console.info('Cloud.save()', this.id)
    if (!items) return
    if (super.save) super.save(items)
    if (networkable.includes(this.type)) this.to_network(items.outerHTML)
  }
  async delete () {
    console.info('Cloud.delete()', this.id)
    const user = firebase.auth().currentUser
    if (user && navigator.onLine) {
      del(this.id)
      await firebase.storage().ref().child(as_filename(this.id)).delete()
    }
  }
}
export default Cloud
