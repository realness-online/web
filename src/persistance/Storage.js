/** @typedef {import('@/types').Item_Id} Item_Id */

// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Local from '@/persistance/Local'
import Large from '@/persistance/Large'
import Cloud from '@/persistance/Cloud'
import Paged from '@/persistance/Paged'
import { current_user, upload } from '@/use/serverless'
import { as_type, as_filename } from '@/utils/itemid'
import { get } from 'idb-keyval'
import { prepare_upload_html } from '@/utils/upload_processor'

export default class Storage {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    this.id = itemid
    this.type = as_type(itemid)
    this.metadata = { contentType: 'text/html' }
  }
}

/** @extends {Storage} */
export class Poster extends Large(Cloud(Storage)) {}

/** @extends {Storage} */
export class Me extends Cloud(Local(Storage)) {
  constructor() {
    super(localStorage.me)
  }
}

/** @extends {Storage} */
export class Relation extends Local(Storage) {
  constructor() {
    super(`${localStorage.me}/relations`)
  }
}

/** @extends {Storage} */
export class Statement extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/statements`)
  }
}

/** @extends {Storage} */
export class Event extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/events`)
  }
}

/** @extends {Storage} */
export class Offline extends Cloud(Storage) {
  async save() {
    const outerHTML = await get(this.id)
    if (outerHTML) await super.save({ outerHTML })
  }
}

/** @extends {Storage} */
export class History extends Cloud(Storage) {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    super(itemid)
  }

  async save(items) {
    // on purpose doesn't call super.save
    if (!items) return false
    if (current_user.value && navigator.onLine) {
      const { compressed, metadata } = await prepare_upload_html(items)
      try {
        await upload(as_filename(this.id), compressed, metadata)
      } catch (e) {
        console.error(e)
        return false
      }
      return true
    }
    return false
  }
}
