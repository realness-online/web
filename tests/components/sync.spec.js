import { vi } from 'vitest'

vi.mock('@/persistance/Storage', () => ({
  Storage: class {
    constructor(itemid) {
      this.id = itemid
    }
  },
  Statements: class {
    async save() {}
  },
  Me: class {
    async save() {}
  },
  Poster: class {
    async save() {}
  },
  Offline: class {
    constructor(itemid) {
      this.id = itemid
    }
    async save() {
      const outerHTML = await get(this.id)
      if (!outerHTML) return

      if (this.id.startsWith('/+/')) {
        const path_parts = this.id.split('/')
        path_parts[1] = localStorage.me.replace('/', '')
        this.id = path_parts.join('/')
      }

      await super.save({ outerHTML })
    }
  }
}))

vi.mock('@/persistance/Cloud', () => ({
  Cloud: Base =>
    class extends Base {
      async save() {}
    }
}))

// Now import everything else
import { shallowMount, mount, flushPromises as flush } from '@vue/test-utils'
import { get, set } from 'idb-keyval'
import * as itemid from '@/utils/itemid'
import * as sync_worker from '@/use/sync'
import sync from '@/components/sync'
import { Storage, Statements, Me, Poster, Offline } from '@/persistance/Storage'
import { Cloud } from '@/persistance/Cloud'
import {
  setup_current_user,
  clear_current_user,
  test_user
} from '../mocks/helpers/user'

const fake_props = {
  global: {
    stubs: ['router-link', 'router-view']
  },
  props: { config: {} }
}
const statement = {
  id: '/+16282281824',
  type: 'statements',
  statement: 'I like to move it'
}
let current_user = {
  phoneNumber: '+16282281824'
}
describe('@/components/sync', () => {
  let wrapper
  let person

  beforeEach(async () => {
    setup_current_user()
    person = {
      first_name: 'Scott',
      last_name: 'Fryxell',
      id: '/+16282281824',
      avatar: '/+16282281824/avatars/1578929551564',
      visited: '2020-03-03T17:37:22.943Z'
    }
    wrapper = shallowMount(sync)
    await flush()
  })

  afterEach(() => {
    clear_current_user()
    vi.clearAllMocks()
  })

  describe('Render', () => {
    it('Renders sync component', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })
  describe('Watchers', () => {
    describe('statement', () => {
      it('Does nothing unless there is a statement', async () => {
        wrapper.setProps({ statement })
        const save_spy = vi.spyOn(Statements.prototype, 'save')
        wrapper.setProps({ statement: null })
        expect(save_spy).not.toBeCalled()
      })
      it('Triggered when statement is set', async () => {
        await wrapper.setProps({ statement })
        await flush()
        expect(wrapper.emitted('update:statement')).toBeTruthy()
      })
    })
    describe('person', () => {
      it('Does nothing unless there is a person', async () => {
        const props = { ...fake_props }
        props.props.person = person
        wrapper = await shallowMount(sync, props)
        wrapper.setProps({ person: null })
        const save_spy = vi.spyOn(Me.prototype, 'save')
        expect(wrapper.emitted('update:person')).not.toBeTruthy()
        expect(save_spy).not.toBeCalled()
      })
      it('Triggered when statement is set', async () => {
        localStorage.sync_time = Date.now()
        const save_spy = vi.spyOn(Me.prototype, 'save')
        wrapper = shallowMount(sync, fake_props)
        await wrapper.setProps({ person })
        await flush()
        expect(save_spy).toBeCalled()
      })
    })
  })
  describe('Methods', () => {
    describe('#visibility_change', () => {
      it('Plays the sync when app becomes visible', async () => {
        // being explicit
        const play = vi
          .spyOn(wrapper.vm, 'play')
          .mockImplementation(() => Promise.resolve())
        Object.defineProperty(window.document, 'visibilityState', {
          enumerable: false,
          configurable: false,
          writable: true,
          value: 'visible'
        })
        await wrapper.vm.play()
        expect(wrapper.vm.play).toBeCalled()
      })
      it('Is chill when the UI is hidden', async () => {
        wrapper = await shallowMount(sync, fake_props)
        wrapper.vm.syncer = {
          postMessage: vi.fn()
        }
        Object.defineProperty(window.document, 'visibilityState', {
          value: 'hidden'
        })
        const play = vi
          .spyOn(wrapper.vm, 'play')
          .mockImplementation(() => Promise.resolve())
        await wrapper.vm.visibility_change()
        expect(document.visibilityState).toBe('hidden')
        expect(play).not.toBeCalled()
      })
    })
    describe('#get_my_itemid', () => {
      it('Returns the user itemid when called without type', async () => {
        expect(localStorage.me).toBe(`/${current_user.phoneNumber}`)
        wrapper = shallowMount(sync, fake_props)
        await flush()
        expect(wrapper.vm.get_my_itemid()).toBe('/+16282281824')
      })
    })
    describe('#sync_offline_actions', () => {
      beforeEach(async () => {
        // Mock online status
        Object.defineProperty(navigator, 'onLine', {
          value: true,
          writable: true
        })
      })

      it('Processes anonymous statements from sync:offline queue', async () => {
        const anonymous_statement = {
          id: '/+/statements/123',
          action: 'save',
          type: 'statement',
          statement: 'Test anonymous statement'
        }
        await set('sync:offline', [anonymous_statement])

        const offline_save_spy = vi.spyOn(Offline.prototype, 'save')
        await sync_worker.sync_offline_actions()

        expect(offline_save_spy).toHaveBeenCalledWith()
        expect(await get('sync:offline')).toBeUndefined()
      })

      it('Migrates anonymous posters using Offline class', async () => {
        const created_at = Date.now()
        const anonymous_poster = {
          type: 'poster',
          content: 'Test poster content'
        }

        // Set up anonymous poster in IndexedDB
        await set('/+/posters/', { items: [created_at] })
        await set(`/+/posters/${created_at}`, anonymous_poster)

        const offline_save_spy = vi.spyOn(Offline.prototype, 'save')
        await sync_worker.sync_offline_actions()

        // Verify poster was migrated using Offline class and cleanup occurred
        expect(offline_save_spy).toHaveBeenCalled()
        expect(await get('/+/posters/')).toBeUndefined()
        expect(await get(`/+/posters/${created_at}`)).toBeUndefined()
      })

      it('Does nothing when offline', async () => {
        Object.defineProperty(navigator, 'onLine', { value: false })

        const offline_save_spy = vi.spyOn(Offline.prototype, 'save')

        await sync_worker.sync_offline_actions()

        expect(offline_save_spy).not.toHaveBeenCalled()
      })
    })
  })
})

describe('@/use/sync', () => {
  let wrapper
  let person

  beforeEach(() => {
    setup_current_user()
    person = {
      first_name: 'Scott',
      last_name: 'Fryxell',
      id: '/+14151234356',
      avatar: '/+14151234356/avatars/1578929551564',
      visited: '2021-02-04T18:30:01.038Z'
    }
    wrapper = shallowMount(vector_mock)
  })

  afterEach(() => {
    clear_current_user()
    vi.clearAllMocks()
  })

  describe('methods', () => {
    describe('#visit', () => {
      it('Exists', () => {
        expect(wrapper.vm.update_visit).toBeDefined()
      })

      it('Updates the user with a visit', async () => {
        const load_spy = vi
          .spyOn(itemid, 'load')
          .mockImplementation(() => Promise.resolve(person))
        await wrapper.vm.update_visit()
        await flushPromises()
        expect(load_spy).toBeCalled()
        expect(wrapper.emitted('update:person')).toBeTruthy()
      })

      it('Does nothing unless there is a person', async () => {
        const load_spy = vi
          .spyOn(itemid, 'load')
          .mockImplementation(() => Promise.resolve(null))
        await wrapper.vm.update_visit()
        await flushPromises()
        expect(load_spy).toBeCalled()
        expect(wrapper.emitted('update:person')).not.toBeTruthy()
      })

      it('Waits a proper interval before update to visit', async () => {
        person.visited = new Date().toISOString()
        const load_spy = vi
          .spyOn(itemid, 'load')
          .mockImplementation(() => Promise.resolve(person))
        await wrapper.vm.update_visit()
        await flushPromises()
        expect(load_spy).toBeCalled()
        expect(wrapper.emitted('update:person')).not.toBeTruthy()
      })
    })
  })
})

describe('Offline Storage', () => {
  let cloud_save_spy

  beforeEach(() => {
    localStorage.me = '/+16282281824'
    cloud_save_spy = vi.spyOn(Cloud(Storage).prototype, 'save')
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  it('Transforms anonymous statement IDs to user IDs', async () => {
    const anonymous_id = '/+/statements/123'
    const expected_id = '/+16282281824/statements/123'
    const statement_html = '<div>Test statement</div>'

    await set(anonymous_id, statement_html)
    const offline = new Offline(anonymous_id)

    await offline.save()

    expect(offline.id).toBe(expected_id)
    expect(cloud_save_spy).toHaveBeenCalledWith({ outerHTML: statement_html })
  })

  it('Transforms anonymous poster IDs to user IDs', async () => {
    const created_at = '1234567890'
    const anonymous_id = `/+/posters/${created_at}`
    const expected_id = `/+16282281824/posters/${created_at}`
    const poster_html = '<div>Test poster</div>'

    await set(anonymous_id, poster_html)
    const offline = new Offline(anonymous_id)

    await offline.save()

    expect(offline.id).toBe(expected_id)
    expect(cloud_save_spy).toHaveBeenCalledWith({ outerHTML: poster_html })
  })

  it('Does not transform already transformed IDs', async () => {
    const user_id = '/+16282281824/statements/123'
    const statement_html = '<div>Already transformed</div>'

    await set(user_id, statement_html)
    const offline = new Offline(user_id)

    await offline.save()

    expect(offline.id).toBe(user_id)
    expect(cloud_save_spy).toHaveBeenCalledWith({ outerHTML: statement_html })
  })

  it('Does nothing when content is not found', async () => {
    const anonymous_id = '/+/statements/not-found'
    const offline = new Offline(anonymous_id)

    await offline.save()

    expect(cloud_save_spy).not.toHaveBeenCalled()
  })
})
