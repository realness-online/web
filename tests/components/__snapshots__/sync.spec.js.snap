// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/components/sync Render Renders sync component 1`] = `
<div
  config="[object Object]"
  hidden=""
>
  <!--v-if-->
  <as-days-stub
    avatars=""
    events=""
    itemid="/+16282281824/statements"
    itemscope=""
    paginate="false"
    posters=""
    statements=""
    working="false"
  />
  <!--v-if-->
  <!--v-if-->
</div>
`;
